delivery_auto_refresh/models/sale_order.py

```diff
@@ -12,14 +12,33 @@ class SaleOrder(models.Model):
     carrier_id = fields.Many2one(
         compute="_compute_carrier_id", store=True, readonly=False
     )
+    available_carrier_ids = fields.Many2many(
+        comodel_name="delivery.carrier", compute="_compute_available_carrier_ids",
+    )
 
     @api.depends("partner_id", "partner_shipping_id")
     def _compute_carrier_id(self):
         if hasattr(super(), "_compute_carrier_id"):
             super()._compute_carrier_id()
         for order in self:
             action = order.action_open_delivery_wizard()
-            order.carrier_id = action["context"]["default_carrier_id"]
+            carrier_id = self.env["delivery.carrier"].browse(
+                action["context"]["default_carrier_id"]
+            )
+            # If the carrier isn't allowed for the current shipping address, we wont
+            # default to it. In that case we'd try to fallback to the former carrier.
+            order.carrier_id = fields.first(
+                (carrier_id | order.carrier_id).filtered(
+                    lambda x: x in order.available_carrier_ids._origin
+                )
+            )
```

Comment:
This seems a bit convoluted, and it can trigger other onchanges with no real change, so please change it to this:

```suggestion
            # If the carrier isn't allowed for the current shipping address, we wont
            # default to it. In that case we'd try to fallback to the former carrier.
            carrier = self.env["delivery.carrier"].browse(
                action["context"]["default_carrier_id"]
            ).filtered(lambda x: x in order.available_carrier_ids._origin)
            if carrier:
                order.carrier_id = carrier
```

Reply:
The problem is that you're not testing if the carrier is allowed for the current shipping address. Case:

- Delivery A - restricted to Spain


Customer commercial partner: delivery method -> Delivery A
Child delivery address in Spain (no delivery method configured)
Child delivery address in France (no delivery method configured)

- We set the order with delivery address in Spain -> Delivery method will be A
- We change the address to France -> Delivery method stays on A which is not correct.

Reply:
OK, got it.